spec/spec.tex

```diff
@@ -977,11 +977,17 @@ \subsection{Instances}\label{instances}
 
 To disallow infinitely recursive hardware, modules cannot contain instances of itself, either directly, or indirectly through instances of other modules it instantiates.
 
-\subsection{Stops}
+\subsection{Stops} \label{stop_stmt}
 The stop statement is used to halt simulations of the circuit. Backends are free to generate hardware to stop a running circuit for the purpose of debugging, but this is not required by the FIRRTL specification.
 
 A stop statement requires a clock signal, a halt condition signal that has a single bit unsigned integer type, and an integer exit code.
 
+For clocked statements that have side effects in the environment (stop, print, and verification
+statements), the order of execution of any such statements that are triggered on the same clock edge
+is determined by their syntactic order in the enclosing module. The order of execution of clocked,
+side-effect-having statements in different modules or with different clocks that trigger
+concurrently is undefined.
+
```

Comment:
This makes sense to me, but I'd like @azidar (on vacation so we'll have to wait) and @aswaterman to take a look.

I believe we do preserve the order of all simulation statements in the implementation, but we need to make sure we're doing this.

Reply:
The rationale is that the current (Chisel 3.3 + FIRRTL 1.3) implementation of asserts _already assumes_ that at least a partial ordering of simulation statements is maintained. Since there is no actual coupling of those two statements (especially after ExpandWhens), this feels like the weakest simple thing that will guarantee that. The idea here is that it's better have a violation of an explicit rule than an implicit assumption.

One could interpret the following sentence from section 5.3.1 in the spec as already implying the ordering from this addition, but the fact that it's listed with connection order makes it less clear than it could be.
> Ordering of statements is significant in a statement group. Intuitively, during elaboration, statements execute in order, and the effects of later statements take precedence over earlier ones.

Reply:
But I am certainly in no hurry to merge this, so whoever can take a look at it, that would be great.

Reply:
@albert-magyar Do I understand this correctly in that if you have a print, a stop and another print that are all triggered together, we should see the output of the first print, but never the output of the second print since the simulation has been stopped?

Reply:
If they appear in that order, yes.